File: java/Samples/BiometricStandards/VB/ANTemplateSampleVB/CreateANPenVectorArrayForm.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Windows.Forms
Imports Neurotec.Biometrics.Standards

Partial Public Class CreateANPenVectorArrayForm
	Inherits Form
	#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
	End Sub

	#End Region

	#Region "Private fields"

	Private _vectors() As ANPenVector

	#End Region

	#Region "Public constructors"

	Public Property Vectors() As ANPenVector()
		Get
			Return _vectors
		End Get
		Set(ByVal value As ANPenVector())
			_vectors = value
			dataGridView.Rows.Clear()
			For Each item As ANPenVector In Vectors
				dataGridView.Rows.Add(item.X, item.Y, item.Pressure)
			Next item
		End Set
	End Property

	#End Region

	#Region "Private form events"

	Private Sub DataGridViewCellEndEdit(ByVal sender As Object, ByVal e As DataGridViewCellEventArgs) Handles dataGridView.CellEndEdit
		Dim row As DataGridViewRow = dataGridView.Rows(e.RowIndex)
		If row.IsNewRow Then
			Return
		End If
		Dim cell As DataGridViewCell = row.Cells(e.ColumnIndex)
		Dim cellValue As Object = cell.FormattedValue
		Dim isValid As Boolean = cellValue IsNot Nothing
		If isValid Then
			Dim isPressure As Boolean = e.ColumnIndex = 2
			If isPressure Then
				Dim value As Byte
				isValid = Byte.TryParse(cellValue.ToString(), value)
			Else
				Dim value As UShort
				isValid = UShort.TryParse(cellValue.ToString(), value)
			End If
		End If

		cell.Style.BackColor = CType(IIf(isValid, Color.White, Color.Red), Color)
	End Sub

	Private Sub BtnOkClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnOk.Click
		Dim isOk As Boolean = True
		Dim list As New List(Of ANPenVector)()
		For Each row As DataGridViewRow In dataGridView.Rows
			If row.IsNewRow Then
				Continue For
			End If

			Dim x As UShort = 0, y As UShort = 0
			Dim pressure As Byte = 0
			If (Not GetValue(row.Cells(0).Value, x)) Then
				isOk = False
				row.Cells(0).Style.BackColor = Color.Red
			End If
			If (Not GetValue(row.Cells(1).Value, y)) Then
				isOk = False
				row.Cells(1).Style.BackColor = Color.Red
			End If
			If (Not GetValue(row.Cells(2).Value, pressure)) Then
				isOk = False
				row.Cells(2).Style.BackColor = Color.Red
			End If
			If (Not isOk) Then
				Exit For
			End If
			list.Add(New ANPenVector(x, y, pressure))
		Next row

		If isOk Then
			Vectors = list.ToArray()
			DialogResult = Windows.Forms.DialogResult.OK
		End If
	End Sub

	#End Region

	#Region "Private methods"

	Private Function GetValue(ByVal cellValue As Object, <System.Runtime.InteropServices.Out()> ByRef value As UShort) As Boolean
		value = 0
		If cellValue Is Nothing Then
			Return False
		End If
		Return UShort.TryParse(cellValue.ToString(), value)
	End Function

	Private Function GetValue(ByVal cellValue As Object, <System.Runtime.InteropServices.Out()> ByRef value As Byte) As Boolean
		value = 0
		If cellValue Is Nothing Then
			Return False
		End If
		Return Byte.TryParse(cellValue.ToString(), value)
	End Function

	#End Region
End Class

File: java/Samples/BiometricStandards/CS/ANTemplateSampleCS/CreateANPenVectorArrayForm.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;
using Neurotec.Biometrics.Standards;

namespace Neurotec.Samples
{
	public partial class CreateANPenVectorArrayForm : Form
	{
		#region Public constructor

		public CreateANPenVectorArrayForm()
		{
			InitializeComponent();
		}

		#endregion

		#region Private fields

		private ANPenVector[] _vectors;

		#endregion

		#region Public constructors

		public ANPenVector[] Vectors
		{
			get { return _vectors; }
			set
			{
				_vectors = value;
				dataGridView.Rows.Clear();
				foreach (ANPenVector item in Vectors)
				{
					dataGridView.Rows.Add(item.X, item.Y, item.Pressure);
				}
			}
		}

		#endregion

		#region Private form events

		private void DataGridViewCellEndEdit(object sender, DataGridViewCellEventArgs e)
		{
			DataGridViewRow row = dataGridView.Rows[e.RowIndex];
			if (row.IsNewRow) return;
			DataGridViewCell cell = row.Cells[e.ColumnIndex];
			object cellValue = cell.FormattedValue;
			bool isValid = cellValue != null;
			if (isValid)
			{
				bool isPressure = e.ColumnIndex == 2;
				if (isPressure)
				{
					byte value;
					isValid = byte.TryParse(cellValue.ToString(), out value);
				}
				else
				{
					ushort value;
					isValid = ushort.TryParse(cellValue.ToString(), out value);
				}
			}

			cell.Style.BackColor = isValid ? Color.White : Color.Red;
		}

		private void BtnOkClick(object sender, EventArgs e)
		{
			bool isOk = true;
			List<ANPenVector> list = new List<ANPenVector>();
			foreach (DataGridViewRow row in dataGridView.Rows)
			{
				if (row.IsNewRow) continue;

				ushort x, y;
				byte pressure;
				if (!GetValue(row.Cells[0].Value, out x))
				{
					isOk = false;
					row.Cells[0].Style.BackColor = Color.Red;
				}
				if (!GetValue(row.Cells[1].Value, out y))
				{
					isOk = false;
					row.Cells[1].Style.BackColor = Color.Red;
				}
				if (!GetValue(row.Cells[2].Value, out pressure))
				{
					isOk = false;
					row.Cells[2].Style.BackColor = Color.Red;
				}
				if (!isOk) break;
				list.Add(new ANPenVector(x, y, pressure));
			}

			if (isOk)
			{
				Vectors = list.ToArray();
				DialogResult = DialogResult.OK;
			}
		}

		#endregion

		#region Private methods

		private bool GetValue(object cellValue, out ushort value)
		{
			value = 0;
			if (cellValue == null) return false;
			return ushort.TryParse(cellValue.ToString(), out value);
		}

		private bool GetValue(object cellValue, out byte value)
		{
			value = 0;
			if (cellValue == null) return false;
			return byte.TryParse(cellValue.ToString(), out value);
		}

		#endregion
	}
}
